pkg/descheduler/descheduler.go

```diff
@@ -215,6 +217,129 @@ func cachedClient(
+		if err != nil {
+			return err
+		}
+	}
+	loopStartDuration := time.Now()
+	defer metrics.DeschedulerLoopDuration.With(map[string]string{}).Observe(time.Since(loopStartDuration).Seconds())
+
+	// if len is still <= 1 error out
+	if len(nodes) <= 1 {
+		klog.V(1).InfoS("The cluster size is 0 or 1 meaning eviction causes service disruption or degradation. So aborting..")
+		return fmt.Errorf("the cluster size is 0 or 1")
+	}
+
+	var client clientset.Interface
+	// When the dry mode is enable, collect all the relevant objects (mostly pods) under a fake client.
+	// So when evicting pods while running multiple strategies in a row have the cummulative effect
+	// as is when evicting pods for real.
+	if rs.DryRun {
+		klog.V(3).Infof("Building a cached client from the cluster for the dry run")
+		// Create a new cache so we start from scratch without any leftovers
+		fakeClient, err := cachedClient(rs.Client, podLister, nodeLister, namespaceLister, priorityClassLister)
+		if err != nil {
+			return err
+		}
+
+		// create a new instance of the shared informer factor from the cached client
+		fakeSharedInformerFactory := informers.NewSharedInformerFactory(fakeClient, 0)
+		// register the pod informer, otherwise it will not get running
+		getPodsAssignedToNode, err = podutil.BuildGetPodsAssignedToNodeFunc(fakeSharedInformerFactory.Core().V1().Pods().Informer())
+		if err != nil {
+			return fmt.Errorf("build get pods assigned to node function error: %v", err)
+		}
+
+		fakeCtx, cncl := context.WithCancel(context.TODO())
+		defer cncl()
+		fakeSharedInformerFactory.Start(fakeCtx.Done())
+		fakeSharedInformerFactory.WaitForCacheSync(fakeCtx.Done())
+
+		client = fakeClient
+		cycleSharedInformerFactory = fakeSharedInformerFactory
+	} else {
+		client = rs.Client
+	}
+
+	klog.V(3).Infof("Building a pod evictor")
+	podEvictor := evictions.NewPodEvictor(
+		client,
+		evictionPolicyGroupVersion,
+		rs.DryRun,
+		deschedulerPolicy.MaxNoOfPodsToEvictPerNode,
+		deschedulerPolicy.MaxNoOfPodsToEvictPerNamespace,
+		nodes,
+		!rs.DisableMetrics,
+		eventRecorder,
+	)
+
+	RunProfiles(ctx, nodes, deschedulerPolicy, client, cycleSharedInformerFactory, podEvictor, getPodsAssignedToNode)
+
+	klog.V(1).InfoS("Number of evicted pods", "totalEvicted", podEvictor.TotalEvicted())
+
+	return nil
+}
+
+// RunProfiles runs all the deschedule plugins of all profiles and
+// later runs through all balance plugins of all profiles. (All Balance plugins should come after all Deschedule plugins)
+// see https://github.com/kubernetes-sigs/descheduler/issues/979
+func RunProfiles(ctx context.Context, nodes []*v1.Node, deschedulerPolicy *api.DeschedulerPolicy, client clientset.Interface,
+	cycleSharedInformerFactory informers.SharedInformerFactory, podEvictor *evictions.PodEvictor, getPodsAssignedToNode podutil.GetPodsAssignedToNodeFunc,
+) {
+	for _, profile := range deschedulerPolicy.Profiles {
+		currProfile, err := frameworkprofile.NewProfile(
```

Comment:
Worth reusing already initialized profile. A profile is currently only a wrapper around the client, pod evictor, informers, etc.

Reply:
initialized profiles separately and used them